### PR TITLE
Roe 96 add page data to session

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link

https://companieshouse.atlassian.net/browse/ROE-96
https://companieshouse.atlassian.net/browse/ROE-97

### Change description

Add page data to the session when posting and 
Get data from the session and render on the page 

**NB: Would like to keep the date as it for now, that is to say with the seperate fields and follow up with an investigation on different ways to handle the data perhaps on another ticket.**

### Work checklist

- [x] Tests added where applicable
- [x] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/model/other.model.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { Address, InputDate, natureOfControl, yesNoResponse } from "./data.types.model";
```
Comment (markpit) [inline src/model/other.model.ts:1]: the name of 'other.ts' is out of sync with naming convention, use the controller/page name ie. beneficial.owner.other.model.ts

File: src/model/application.model.ts [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 export const APPLICATION_DATA_KEY = 'roe';
 
 export interface ApplicationData {
     presenter?: presenterType.Presenter;
     entity?: entityType.Entity;
     beneficialOwnerType?: beneficialOwnerTypeType.BeneficialOwnerType;
+    beneficialOwnerOther?: beneficialOwnerOtherType.BeneficialOwnerOther;
 }
 
 export type ApplicationDataType = presenterType.Presenter | entityType.Entity | dataType.Address | beneficialOwnerTypeType.BeneficialOwnerType;
```
Comment (mwestacott) [inline src/model/application.model.ts:12]: Don't fully understand what this does, but... do you not need to add 'beneficialOwnerOtherType.BeneficialOwnerOther' on to the end here?

If in fact you do (I've just done this in my PR too) I also wonder how the tests pass without it?

File: src/model/data.types.model.ts [hunk 3]
```diff
@@ -16,3 +16,16 @@ export enum BeneficialOwnerTypeChoice {
     otherLegal = "otherLegal",
     none = "none"
 }
+
+
+export interface InputDate {
```
Comment (mwestacott) [inline src/model/data.types.model.ts:21]: Could we not just name this 'Date', or is there a name clash with a reserved word (or something)? All of these things are 'input'....

File: src/model/other.model.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    statement2
+}
+
+export const DateKey: string = "startDate";
+export const DateKeys: string[] = ["startDate-day", "startDate-month", "startDate-year"];
+
+export const BeneficialOwnerOtherKey: string = "beneficialOwnerOther";
+export const BeneficialOwnerOtherKeys: string[] = ["corporationName", "principalAddress", "isSameAddress", "serviceAddress", "lawGoverned", "startDate", "natureOfControl", "statementCondition", "isSactioned"];
+export interface BeneficialOwnerOther {
+    corporationName: string
```
Comment (mwestacott) [inline src/model/other.model.ts:14]: Should more of these fields be optional? i.e. have '?' at the end of the names?

File: views/beneficial-owner-other.html [hunk 5]
```diff
@@ -44,27 +45,30 @@ <h1 class="govuk-heading-xl govuk-!-margin-0">Tell us about the corporate benefi
...
           label: {
             text: "Town or city"
           },
           classes: "govuk-!-width-two-thirds",
-          id: "address-town",
-          name: "address-town",
+          id: "principalAddressTown",
+          name: "principalAddressTown",
+          value: principalAddress.principalAddressTown if principalAddress,
           autocomplete: "address-level2"
```
Comment (mwestacott) [inline views/beneficial-owner-other.html:72]: A general question really (again, also pondered this with my change) - should autocomplete value match the 'name' field? You did this above for address lines1 and 2 but not here. I know that prototype template code seems to be inconsistent here too.

File: src/model/application.model.ts (hunk 2 above)
Comment (markpit) [reply src/model/application.model.ts:12]: prepareData will return data in the 'shape' of ApplicationDataType which I think can be one of those types in the list. But I think at compile time it just looks and sees
`const data: ApplicationDataType = prepareData(req.body, beneficialOwnerTypeType.BeneficialOwnerTypeKeys);`
which satisfies the compiler.
It's only at runtime that the shape of the object returned is defined but I would have thought you'd get an error then but seems to work. So perhaps the ApplicationDataType isn't doing the limiting we hoped, but as all the other types are in that list then this new beneficialOwnerOtherType.BeneficialOwnerOther should be too.

Comment (mwestacott) [reply src/model/application.model.ts:12]: 👍  worth getting @mouhajer-ch 's thoughts on this one too.

File: src/model/data.types.model.ts (hunk 3 above)
Comment (mattch1) [reply src/model/data.types.model.ts:21]: "Date" clashes
We want the name to be as generic as possible.

File: views/beneficial-owner-other.html [hunk 6]
```diff
@@ -183,12 +196,13 @@ <h1 class="govuk-heading-xl govuk-!-margin-0">Tell us about the corporate benefi
...
-          id: "corp-law-governed",
-          name: "corp-law-governed"
+          id: "law-governed",
+          name: "lawGoverned",
+          value: lawGoverned
         }) }}
         {{ govukDateInput({
-          id: "corp-startdate",
-          namePrefix: "corp-startdate",
+          id: "start-date",
```
Comment (mwestacott) [inline views/beneficial-owner-other.html:204]: Should this also be named 'startDate'?

Comment (mattch1) [reply views/beneficial-owner-other.html:204]: I'm just going to leave this as is.

File: views/beneficial-owner-other.html (hunk 5 above)
Comment (mattch1) [reply views/beneficial-owner-other.html]: Updated this.

File: src/model/other.model.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattch1) [reply src/model/other.model.ts]: Made all fields optional.

File: src/model/other.model.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import { Address, InputDate, natureOfControl, yesNoResponse } from "./data.types.model";
+
+export enum statementCondition {
```
[the remark this answers is not in the document]
Comment (mattch1) [reply src/model/other.model.ts]: Updated this.

File: src/model/data.types.model.ts [hunk 8]
```diff
@@ -16,3 +16,16 @@ export enum BeneficialOwnerTypeChoice {
...
+    day: number
+    month: number
+    year: number
+}
+
+export enum natureOfControl {
+    over25under50,
```
[the remark this answers is not in the document]
Comment (mattch1) [reply src/model/data.types.model.ts]: staying with "over" as this is what is on the prototype. However, added percent.

File: test/controllers/beneficial.owner.other.controller.spec.ts [hunk 9]
```diff
@@ -1,32 +1,80 @@
...
+  describe("GET tests", () => {
+
+    test("renders the page through GET", async () => {
+      mockGetApplicationData.mockReturnValueOnce(BENEFICIAL_OWNER_OTHER_OBJECT_MOCK);
+      const resp = await request(app).get(BENEFICIAL_OWNER_OTHER_URL);
+
+      expect(resp.status).toEqual(200);
+      expect(resp.text).toContain(PAGE_TITLE);
+    });
+
```
Comment (mwestacott) [inline test/controllers/beneficial.owner.other.controller.spec.ts:41]: I think we need a test for getting the data out of the session and displaying on page.

File: src/model/data.types.model.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mwestacott) [reply src/model/data.types.model.ts]: Thanks for adding the 'percent'.

Looked at your screen and it matches mine. Options are:

- over 25% up to (and including) 50%
- more than 50% and less than 75%
- 75% or more

In order for this to be reusable and more accurate, I suggest renaming the first and last of these to:

_over25upTo50Percent_ and _75PercentOrMore_.

File: src/model/data.types.model.ts [hunk 10]
```diff
@@ -16,3 +16,21 @@ export enum BeneficialOwnerTypeChoice {
...
+
+
+export interface InputDate {
+    day: number
+    month: number
+    year: number
+}
+
+export enum natureOfControl {
+    over25under50Percent = "25",
```
Comment (mwestacott) [inline src/model/data.types.model.ts:28]: Thanks for adding the 'percent'.

Looked at your screen and it matches mine. Options are:

over 25% up to (and including) 50%
more than 50% and less than 75%
75% or more
In order for this to be reusable and more accurate, I suggest renaming the first and last of these to:

over25upTo50Percent and 75PercentOrMore.

Comment (mattch1) [reply src/model/data.types.model.ts:28]: "atLeast75Percent" would be better I don't really want to start the name with a number.

Comment (mwestacott) [reply src/model/data.types.model.ts:28]: Sounds good.

File: test/controllers/beneficial.owner.other.controller.spec.ts [hunk 11]
```diff
@@ -1,32 +1,80 @@
...
+  beforeEach(() => {
+    jest.clearAllMocks();
   });
 
-  test("posts the page and renders the managing-officer page", async () => {
-    const resp = await request(app).post(BENEFICIAL_OWNER_OTHER_URL);
+  describe("GET tests", () => {
+
+    test("renders the page through GET", async () => {
+      mockGetApplicationData.mockReturnValueOnce(BENEFICIAL_OWNER_OTHER_JSON_MOCK);
```
Comment (markpit) [inline test/controllers/beneficial.owner.other.controller.spec.ts:35]: mockGetApplicationData should return the APPLICATION_DATA_MOCK, you can add the BENEFICIAL_OWNER_OTHER_OBJECT_MOCK to the APPLICATION_DATA_MOCK and then you don't need the BENEFICIAL_OWNER_OTHER_JSON_MOCK.

File: src/model/data.types.model.ts [hunk 12]
```diff
@@ -17,17 +17,16 @@ export enum BeneficialOwnerTypeChoice {
...
     month: number
     year: number
 }
 
 export enum natureOfControl {
-    over25under50Percent = "25",
-    over50under75Percent = "50",
-    over75Percent = "75"
+    over25upTo50Percent = "25",
+    over50upTo75Percent = "50",
```
Comment (mwestacott) [inline src/model/data.types.model.ts:28]: Naming of second one here was fine - "more than 50% and less than 75%" - but now name suggests a value of 75% is allowed. Suggest changing it back to what you had: _over50under75Percent_.

File: views/beneficial-owner-other.html [hunk 13]
```diff
@@ -162,7 +162,7 @@ <h1 class="govuk-heading-xl govuk-!-margin-0">Tell us about the corporate benefi
 
         {{ govukRadios({
           classes: "govuk-radios",
-          idPrefix: "same-address",
+          idPrefix: "issameAddress",
```
Comment (mwestacott) [inline views/beneficial-owner-other.html:165]: Typo, well case. isSameAddress.

File: views/beneficial-owner-other.html [hunk 14]
```diff
@@ -288,7 +288,7 @@ <h1 class="govuk-heading-xl govuk-!-margin-0">Tell us about the corporate benefi
 
         {{ govukRadios({
           classes: "govuk-radios",
-          idPrefix: "is-sactioned",
+          idPrefix: "isSactioned",
```
Comment (mwestacott) [inline views/beneficial-owner-other.html:291]: Surely 'is sanctioned' is meant here and we should correct typo in the names?

**Review by markpit — APPROVED** [review]

**Pull request merged**